### PR TITLE
Support tracers in `PerturbationAdvection` open boundary conditions

### PR DESCRIPTION
**Pull request by tomchor**
## Summary

When https://github.com/CliMA/Oceananigans.jl/pull/5314 introduced the possibility to use `PerturbationAdvection` tracers, it didn't actually change the calculation of the advection velocity.

This PR takes a closer look at that and fixes two independent problems.

### 1. Advecting velocity

The radiation phase speed was computed from the field's *own* exterior value:

```julia
c = ψ̄ + c★
```

For a boundary-normal velocity component, `ψ̄` *is* the advecting velocity, so this is correct. For a tracer, `ψ̄` is the tracer value (e.g. a temperature), and using it as an advection speed is physically meaningless — the inflow/outflow timescale selection (`ψ̄ >= 0`) was likewise testing the sign of the *tracer* rather than of the *flow*.

The scheme now takes the **boundary-normal flow velocity** as the phase speed. `step_right/left_open_boundary!` accept an explicit advecting velocity:

- velocity (`Face`) halo methods pass `nothing` and fall back to `ψ̄` → **momentum behavior is byte-for-byte unchanged**;
- tracer (`Center`) halo methods pass the normal velocity component (`u`/`v`/`w`) read from `model_fields` at the boundary face (no transverse interpolation needed — the normal velocity is already center-located in the transverse directions).

### 2. Vertical (top/bottom) boundaries

`_fill_top_halo!` / `_fill_bottom_halo!` only had methods for `Face`-located fields. A tracer with a `PerturbationAdvection` BC on `top`/`bottom` silently fell through to the generic open-BC halo fill (a plain prescribed-value condition). Added the missing `Center` methods.

## Tests

Added `test_perturbation_advection_tracer_open_boundary_conditions` to `test/test_boundary_conditions_integration.jl`, run alongside the existing velocity test in the `Open boundary conditions` testset. It uses `HydrostaticFreeSurfaceModel` (the model that fills tracer open boundaries):

- **Lateral (east/west):** a uniform flow makes one x-boundary inflow and the other outflow. Inflow relaxes instantly to the exterior value; outflow radiates and preserves the uniform interior. Same tracer config, opposite velocity → opposite results — the case that fails on the old code.
- **Vertical (top/bottom):** in a quiescent column the scheme freezes the boundary, so a uniform tracer stays uniform — confirming the new `Center` top/bottom methods run the radiation scheme instead of the plain value-BC fallback.

Verified locally (CPU): the new tracer test and the existing velocity regression test both pass (15/15 `@test`s). `test_quality_assurance.jl` left to CI.

## Scope / follow-ups

Deliberately **not** in this PR:

- **West/south/bottom `Center` index asymmetry** — the low-side `Center` halo methods write the first interior cell rather than the halo cell. Pre-existing; should be addressed separately.
- **`NonhydrostaticModel` plumbing** — tracer halos are filled there with `fill_open_bcs=false`, so tracer open BCs are currently inert in the nonhydrostatic model. `HydrostaticFreeSurfaceModel` already fills them (`fill_open_bcs=true`). Enabling the nonhydrostatic path is a time-stepping change for a separate PR.
- **Density-weighted advecting velocity** — for anelastic models with prognostic `ρu`, the extracted `U` would be momentum, not velocity; converting it via `to_intensive` is left as follow-up.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: test/test_boundary_conditions_integration.jl [hunk 1]
```diff
@@ -156,6 +156,68 @@ function test_perturbation_advection_open_boundary_conditions(arch, FT)
     end
 end
 
+function test_perturbation_advection_tracer_open_boundary_conditions(arch, FT)
+    # Tracer (Center-located) open boundaries are filled within the
+    # HydrostaticFreeSurfaceModel update, so the scheme is exercised there.
+    grid = RectilinearGrid(arch, FT; topology = (Bounded, Flat, Bounded),
+                           size = (4, 4), x = (0, 4), z = (0, 4))
```
[the remark this answers is not in the document]
Comment (tomchor) [reply test/test_boundary_conditions_integration.jl:163]: Wrong

**Comment by tomchor** [comment]
Here's a simulation of a tracer disturbance using a uniform, fixed velocity in three cases, each with a different background value. The external value for the tracers (used by `OpenBoundaryCondition`) is the same as the background value, and I'm using a `Centered(order=4)` advection scheme since it illustrates the difference better (hence the grid-scale noise).

On `main`:

https://github.com/user-attachments/assets/57ec2809-4d63-48df-ba6a-4315ae589c43

On this PR:

https://github.com/user-attachments/assets/bf5ec848-833a-441b-ab71-ed1c25e31481

File: src/BoundaryConditions/perturbation_advection.jl [hunk 2]
```diff
@@ -152,12 +158,15 @@ const PAOBC = BoundaryCondition{<:Open{<:PerturbationAdvection}}
...
+    Uᵃ = advecting_velocity(U, ψ̄)
+
+    # Phase speed: advecting velocity + gravity wave speed
+    c = Uᵃ + c★
     Ũ = max(0, min(1, Δt / ΔX * c))
 
     # Inflow vs outflow relaxation
-    τ = ifelse(ψ̄ >= 0, pa.outflow_timescale, pa.inflow_timescale)
+    τ = ifelse(Uᵃ >= 0, pa.outflow_timescale, pa.inflow_timescale)
     τ̃ = Δt / τ
```
[the remark this answers is not in the document]
Comment (tomchor) [reply src/BoundaryConditions/perturbation_advection.jl:170]: > *(reply drafted by Claude)*

Agreed on the diagnosis — but three reasons I'd keep this as a follow-up rather than land a fix in this PR:

1. **Already disclosed.** The PR description lists this under *Scope / follow-ups*:
   > *Density-weighted advecting velocity — for anelastic models with prognostic `ρu`, the extracted `U` would be momentum, not velocity; converting it via `to_intensive` is left as follow-up.*

2. **The right fix isn't obvious.** `to_intensive(ρ, U, k)` is correct *if and only if* `model_fields.u` is itself density-weighted. In a regular Boussinesq model where someone sets `density = ρ` to handle a `ρθ`-like tracer but where `u` is still intensive, an unconditional conversion of `U` would *introduce* a bug (divide a velocity by ρ). A proper fix probably needs a separate flag (e.g. `velocity_is_density_weighted`) or a different convention — not a blanket conversion.

3. **No in-tree consumer.** Oceananigans doesn't currently have an anelastic model with prognostic `ρu`; the `density` parameter was added in #5314 for anticipated downstream use. CI doesn't exercise the bug today and wouldn't exercise the fix either, so a change here would land unverified.

Resolving — happy to revisit once an anelastic consumer lands and the convention is pinned down.

File: test/test_boundary_conditions_integration.jl [hunk 3]
```diff
@@ -156,6 +156,108 @@ function test_perturbation_advection_open_boundary_conditions(arch, FT)
...
+
+    Nx, Nz, Hx, Hz = grid.Nx, grid.Nz, grid.Hx, grid.Hz
+
+    c̄ = 1   # prescribed exterior tracer value
+    c₀ = 5  # uniform initial tracer value (interior and halos)
+
+    east_halo(c) = Array(parent(c))[Nx + 1 + Hx, 1, Hz + 1]
+    top_halo(c)  = Array(parent(c))[Hx + 1, 1, Nz + 1 + Hz]
+    low_side(c)  = Array(interior(c))[1, 1, 1]  # west and bottom write the first interior cell
```
Comment (glwagner) [inline test/test_boundary_conditions_integration.jl]:
```suggestion
    low_side(c)  = Array(interior(c, 1, 1, 1))  # west and bottom write the first interior cell
```

does this work?

**Review by glwagner — APPROVED** [review]

File: test/test_boundary_conditions_integration.jl (hunk 3 above)
Comment (glwagner) [reply test/test_boundary_conditions_integration.jl]: I also wonder if `Array(parent(c))[i, j, k]` can be improved. For example `@allowscalar parent(c)[i, j, k]` might be better, less wasteful than sending the whole array to CPU and then just picking one index. and easier to read.

**Comment by jagoosw** [comment]
> Here's a simulation of a tracer disturbance using a uniform, fixed velocity in three cases, each with a different background value. The external value for the tracers (used by `OpenBoundaryCondition`) is the same as the background value, and I'm using a `Centered(order=4)` advection scheme since it illustrates the difference better (hence the grid-scale noise).
> 
> On `main`:
> 
>  tracer_outflow.jld2-c4-main.mp4 
> On this PR:
> 
>  tracer_outflow.jld2-c4-open-tracers.mp4

How well does this example do with just a zero gradient boundary on the exit? I read once that for tracers it makes minimal difference.

**Comment by tomchor** [comment]
> > Here's a simulation of a tracer disturbance using a uniform, fixed velocity in three cases, each with a different background value. The external value for the tracers (used by `OpenBoundaryCondition`) is the same as the background value, and I'm using a `Centered(order=4)` advection scheme since it illustrates the difference better (hence the grid-scale noise).
> > On `main`:
> > tracer_outflow.jld2-c4-main.mp4
> > On this PR:
> > tracer_outflow.jld2-c4-open-tracers.mp4
> 
> How well does this example do with just a zero gradient boundary on the exit? I read once that for tracers it makes minimal difference.

Expectedly it, looks pretty good:

https://github.com/user-attachments/assets/2eee7e92-df1a-4a0a-86eb-ab7caf9c24ce

But the issue is that, in realistic simulations, you want to be able to advect an exterior value in through a boundary. Likewise you also want to advect the interior value out through the same boundary. A radiation boundary is the way to do that. Also I think for active tracers a zero-gradient BC will cause reflections.

File: test/test_boundary_conditions_integration.jl (hunk 3 above)
Comment (tomchor) [reply test/test_boundary_conditions_integration.jl]: It should! Let's try

**Comment by tomchor** [comment]
cc @glwagner @simone-silvestri 

Can someone with force push access merge this? CI is a little finicky atm, so it takes some retries to get tests to pass. That makes CI slow, so every time the tests pass (and they do pass) there's another update to `main` and then I have to re-run the CI again and I never get to merge. Tests are passing though, so this is good to merge.
